pkg/rcmgr/resource.go

```diff
@@ -173,11 +176,92 @@ func (rc *resources) removeConns(incount, outcount, fdcount int) {
 	}
 }
 
+func (rc *resources) addTask(num int, prio ReserveTaskPriority) error {
+	if prio == ReserveTaskPriorityHigh {
+		return rc.addTasks(num, 0, 0)
+	} else if prio == ReserveTaskPriorityMedium {
+		return rc.addTasks(0, num, 0)
+	} else {
+		return rc.addTasks(0, 0, num)
+	}
+}
+
+func (rc *resources) addTasks(high, medium, low int) error {
+	if rc.ntasksHigh+rc.ntasksMedium+rc.ntasksLow+high+medium+low > rc.limit.GetTaskTotalLimit() {
+		return &ErrTaskLimitExceeded{
+			current:   rc.ntasksHigh + rc.ntasksMedium + rc.ntasksLow,
+			attempted: high + medium + low,
+			limit:     rc.limit.GetTaskTotalLimit(),
+			err:       fmt.Errorf("total task limit exceeded: %w", ErrResourceLimitExceeded),
+		}
+	}
+	if high+rc.ntasksHigh > rc.limit.GetTaskLimit(ReserveTaskPriorityHigh) {
+		return &ErrTaskLimitExceeded{
+			current:   rc.ntasksHigh,
+			attempted: high,
+			limit:     rc.limit.GetTaskLimit(ReserveTaskPriorityHigh),
+			err:       fmt.Errorf("high priority task limit exceeded: %w", ErrResourceLimitExceeded),
+		}
+	}
+	if medium+rc.ntasksMedium > rc.limit.GetTaskLimit(ReserveTaskPriorityMedium) {
+		return &ErrTaskLimitExceeded{
+			current:   rc.ntasksMedium,
+			attempted: medium,
+			limit:     rc.limit.GetTaskLimit(ReserveTaskPriorityMedium),
+			err:       fmt.Errorf("medium priority task limit exceeded: %w", ErrResourceLimitExceeded),
+		}
+	}
+	if low+rc.ntasksLow > rc.limit.GetTaskLimit(ReserveTaskPriorityLow) {
+		return &ErrTaskLimitExceeded{
+			current:   rc.ntasksLow,
+			attempted: low,
+			limit:     rc.limit.GetTaskLimit(ReserveTaskPriorityLow),
+			err:       fmt.Errorf("low priority task limit exceeded: %w", ErrResourceLimitExceeded),
+		}
+	}
+	rc.ntasksHigh += high
```

Comment:
Need to protect with atomic operation

Reply:
This component is not thread-safe, and the mutex is used when it is used externally. External users normally use resourceScope which has the mutex~